File: src/leetcode/easy/Solution69.java
package leetcode.easy;

public class Solution69 {
    public int mySqrt(int x) {
        if (x == 0) {
            return 0;
        }
        int start = 1;
        int end = x;
        while (true) {
            int mid = start + (end - start) / 2;
            if (mid <= x / mid && (mid + 1) > x / (mid + 1)) {
                return mid;
            } else if (mid > x / mid) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
    }
}

File: src/leetcode/easy/Solution69.cs
// https://leetcode.com/problems/sqrtx/

public class Solution {
    public int MySqrt(int x) {
        if(x == 0)
        {
            return 0;
        }
        
        int start = 1;
        int end = x/2;

        while(end >= start)
        {
            int mid = start + (end - start)/2;
            long sqrt = (long)mid * mid;

            if(sqrt == x)
            {
                return mid;
            }
            else if(sqrt > x)
            {
                end = mid-1;
            }
            else
            {
                start = mid+1;
            }
        }
        return end;
    }
}
